src/webview/html/messageAsHtml.js

```diff
@@ -72,13 +77,87 @@ $!${messageReactionListAsHtml(reactions, ownUser.user_id, allImageEmojiById)}
 `;
 };
 
-const widgetBody = (message: Message | Outbox) => template`
+/**
+ * Render the body of a message that has submessages.
+ *
+ * Must not be called on a message without any submessages.
+ */
+const widgetBody = (message: Message, ownUserId: UserId) => {
+  invariant(
+    message.submessages !== undefined && message.submessages.length > 0,
+    'should have submessages',
+  );
+
+  const widgets = message.submessages
+    .filter(submessage => submessage.msg_type === 'widget')
+    .map(submessage => ({
+      ...submessage,
+      content: JSON.parse(submessage.content),
+    }));
+
+  const poll_widgets = widgets.filter(widget => widget.content.widget_type === 'poll');
+  // "events" are submessages other than the main widget submessage - for
+  // instance, `vote`, `new_option`, and similar submessages. These are not
+  // documented, look at the PollData code if you're interested in what the
+  // possibilities are.
+  const events = widgets.filter(widget => 'type' in widget.content);
+
+  if (poll_widgets.length !== 1) {
```

Comment:
Does this logic correspond to what the webapp does?

I think it'd be best to try to make this logic track what's in the webapp. Particularly in the absence of documentation.

It looks like the corresponding logic here is around `do_process_submessages` in `static/js/submessage.js`, and its callees.

Reply:
That code does this:

```javascript
    // Right now, our only use of submessages is widgets.

    const data = events[0].data;

    if (data === undefined) {
        return;
    }
```

Which seems strictly worse, to me. (It assumes `events[0]` is the widget submessage, and it assumes that there's only one widget in the message).

I guess it'd probably be for the best to do `if (poll_widgets.length == 0)`, so we're less likely to break in a different way from the webapp in the future. But I don't think we should replace the `.filter()` here with `[0]`, for instance.

Reply:
Added the sorting code, thanks for catching that.

> It also wouldn't make any sense for votes to come after the creation of the poll itself.

Presumably you mean before, not after?

> Basically I would much rather be encoding the assumptions that are in the web implementation -- which was written by Steve, who also specced out this subsystem in the first place -- than encoding a different set of implicit assumptions from our own differently-shaped implementation.

I think that this is reasonable in general, be we do actually have a slightly different set of constraints than the webapp — the mobile app needs to be more robust to future server versions, whereas the webapp can safely assume that (for instance), all submessages will be widgets, since the clientside code will change at the same time as the server side code.

I've rewritten the logic to be much more similar to the webapp, but there are still a couple checks that I've added that I think make sense to be there.

Reply:
> Presumably you mean before, not after?

Ah, yep.


> the mobile app needs to be more robust to future server versions, whereas the webapp can safely assume that (for instance), all submessages will be widgets, since the clientside code will change at the same time as the server side code.

True! I'm happy to have checks for things like that, to handle forward-compatibility that the webapp can handle by just changing the client code when the time comes.

The really essential thing I want to avoid here is the opposite direction, where we encode assumptions that aren't in the web implementation nor documented for the API. That's how we wind up with bugs like the one where we [assumed a new message sent by self was never unread](https://chat.zulip.org/#narrow/stream/3-backend/topic/muted.20users/near/1171634).

And then the practical way to accomplish that -- at least for the situation this subsystem is in, where it's impossible to use only assumptions that are documented in the API, because there is no API documentation -- is to keep our implementation's logic generally shaped in a similar way to the logic in the web implementation, so that it's possible to compare without a lot of work.

Another aspect of forward compatibility is that if we encounter something from the server that's outside the current known patterns, it's not enough to just do something; we have to do the right thing, or do an appropriate kind of nothing. If we go and give meanings to a wide swath of currently-unused protocol space, that's just as capable of conflicting with future API extensions as any bad assumption.

Then in order to effectively avoid doing that, what's needed is basically to have some kind of coherent, and more or less explicit, theory of how the API is meant to be extended in the future. That's a key part of what our API types are for. For example the new `WidgetData` type discusses this pretty directly:
```js
/**
 * The data encoded in a submessage to make the message a widget.
 *
 * Note that future server versions might introduce new types of widgets, so
 * `widget_type` could be a value not included here.  But when it is one of
 * these values, the rest of the object will follow this type.
 */
// Ideally we'd be able to express both the known and the unknown widget
// types: we'd have another branch of this union which looked like
//   | {| +widget_type: (string *other than* those above), +extra_data?: { ... } |}
// But there doesn't seem to be a way to express that in Flow.
export type WidgetData =
  | {|
      +widget_type: 'poll',
      +extra_data?: {| +question?: string, +options?: $ReadOnlyArray<string> |},
    |}
  // We can write these down more specifically when we implement these widgets.
  | {| +widget_type: 'todo', +extra_data?: { ... } |}
  | {| +widget_type: 'zform', +extra_data?: { ... } |};
```

Then it's possible to
* look at the code, and compare it to the explicit expectations, and check they're aligned;
* and look at the explicit expectations, and discuss them or think about them from an API perspective and evaluate whether they seem right, without having to reverse-engineer the implementation to extract its assumptions as part of that discussion.

Reply:
Yeah, that all makes sense :+1: